Use api v4 for airbrake deploy notices

phumpal opened the pull request:
##### SUMMARY
The Airbrake v2 deploy endpoint has been [deprecated](https://airbrake.io/docs/legacy-xml-api/#tracking-deploys).

This PR updates the `airbrake_deployment` module to be compatible w/the v4 deploy endpoint

- https://airbrake.io/docs/api/#deploys-v4
- https://airbrake.io/docs/features/deploy-tracking/

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
airbrake_deployment

##### ADDITIONAL INFORMATION
To notify Airbrake of a deploy using the v4 endpoint requires the Airbrake

- project_id
- project_key

and expects `POST` data in json format.

Examples below assume project id and key are valid.

**Before**
```
cat > test_old.yml <<EOF
- hosts: localhost

  tasks:
    - name: notify airbrake about deploy
      airbrake_deployment:
        repo: https://github.com/phumpal/ansible
        token: 'toinfinityandbeyond'
        environment: 'production'
        user: '{{ lookup("env", "DEPLOY_USER") or lookup("env", "USER") }}'
        revision: '012345'
EOF

ansible-playbook -c local test_old.yml -vvvv

changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "environment": "production",
            "repo": "https://github.com/phumpal/ansible",
            "revision": "012345",
            "token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "url": "https://api.airbrake.io/deploys.txt",
            "user": "phumpal",
            "validate_certs": true
        }
    }
}
```
**After**
```
cat > test.yml <<EOF
- hosts: localhost

  tasks:
    - name: notify airbrake about deploy
      airbrake_deployment:
        repo: https://github.com/phumpal/ansible
        project_id: '123456'
        project_key: 'toinfinityandbeyond'
        environment: 'production'
        user: '{{ lookup("env", "DEPLOY_USER") or lookup("env", "USER") }}'
        revision: 'f06196ff4a'
EOF

ansible-playbook -c local test.yml -vvvv

changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "environment": "production",
            "project_id": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "project_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "repo": "https://github.com/phumpal/ansible",
            "revision": "f06196ff4a",
            "url": "https://api.airbrake.io/api/v4/projects/",
            "user": "phumpal",
            "validate_certs": true
        }
    }
}
```

phumpal commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -78,48 +89,52 @@ def main():
+    if info['status'] == 201:
         module.exit_json(changed=True)
     else:
         module.fail_json(msg="HTTP result code: %d connecting to %s" % (info['status'], url))
```
I don't disagree but perhaps we could do one of the following instead

> module.fail_json(msg="HTTP result code: %d connecting to %s" % (info['status'], mode.params.get('url')))

This display the projects URL api w/out parameter and still allows Airbrake compatible servers like errbit to have a decent error message 

or

> module.fail_json(msg="HTTP result code: %d connecting to Airbrake" % (info['status']))

to remain ambiguous. 

Alternatively we could simply say `connecting to API`

felixfontein replied:
Your first option is what @Akasurde proposed (except using `[]` instead of `.get()`). I would go for that one as well.

phumpal replied:
Thanks @felixfontein. Commit pushed.

phumpal commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -16,15 +16,24 @@
-   - Notify airbrake about app deployments (see http://help.airbrake.io/kb/api-2/deploy-tracking)
+   - Notify airbrake about app deployments (see https://airbrake.io/docs/api/#deploys-v4)
 options:
-  token:
```
Added. PTAL

phumpal commented:
Shall I squash?

felixfontein commented:
No need to, it will be squash-merged anyway.

phumpal commented:
Thanks for your help!

phumpal commented:
@felixfontein all tests pass now.

@Akasurde your requested changes were added.

I'll squash and force-push. Thanks for your assistance.

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -45,21 +52,26 @@
     type: bool
+  token:
+     description:
+      - Parameter 'token' (API token) is deprecated in 2.10. Please remove entry.
```
```suggestion
      - This parameter (API token) has been deprecated in Ansible 2.10. Please remove it from your tasks.
```

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -16,14 +16,21 @@
+  project_id:
+    description:
+      - Airbrake PROJECT_ID
+    required: true
```
This change breaks backwards compatibiliy. Existing tasks will suddenly stop working, since they are missing the new options `project_id` and `project_key`. You need to implement some fallback mechanism that keeps existing playbooks working.

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -78,47 +90,55 @@ def main():
         module.exit_json(changed=True)
     else:
-        module.fail_json(msg="HTTP result code: %d connecting to %s" % (info['status'], url))
+        module.fail_json(msg="HTTP result code: %d connecting to %s" % (info['status'], module.params['']))
```
I think you are missing something. There is no parameter `''` of the module.

phumpal commented:
@felixfontein I've made the unfortunate mistake of rebasing master If you could assist.

phumpal commented:
I've pushed a fix to maintain backwards compatibility with "token" based auth as well as legacy parameters.

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -78,44 +90,78 @@ def main():
+            url=dict(required=False, default='https://api.airbrake.io/api/v4/projects/'),
             validate_certs=dict(default='yes', type='bool'),
         ),
         supports_check_mode=True
```
```suggestion
        supports_check_mode=True,
        required_together=[('project_id', 'project_key')],
        mutually_exclusive=[('project_id', 'token')],
```

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -78,44 +90,78 @@ def main():
+            params["deploy[scm_revision]"] = module.params["revision"]
 
-    url = module.params.get('url')
+        module.deprecate("Parameter 'token' is deprecated in 2.10. Please remove entry.", '2.14')
```
```suggestion
        module.deprecate("Parameter 'token' is deprecated in 2.10. Please remove it and use 'project_id' and 'project_key' instead", version='2.14')
```

felixfontein commented on tests/sanity/ignore-2.10.txt:
```diff
@@ -945,7 +945,6 @@ plugins/modules/identity/keycloak/keycloak_clienttemplate.py validate-modules:pa
 plugins/modules/identity/opendj/opendj_backendprop.py validate-modules:doc-missing-type
 plugins/modules/identity/opendj/opendj_backendprop.py validate-modules:parameter-type-not-in-doc
-plugins/modules/monitoring/airbrake_deployment.py validate-modules:doc-default-does-not-match-spec
 plugins/modules/monitoring/airbrake_deployment.py validate-modules:doc-missing-type
```
Would you mind fixing this one as well? It is about missing `type:` in option docs.

phumpal replied:
Done

felixfontein commented on tests/sanity/ignore-2.9.txt:
```diff
@@ -967,7 +967,6 @@ plugins/modules/messaging/rabbitmq/rabbitmq_user.py validate-modules:parameter-l
 plugins/modules/messaging/rabbitmq/rabbitmq_vhost.py validate-modules:doc-missing-type
 plugins/modules/messaging/rabbitmq/rabbitmq_vhost_limits.py validate-modules:parameter-type-not-in-doc
-plugins/modules/monitoring/airbrake_deployment.py validate-modules:doc-default-does-not-match-spec
 plugins/modules/monitoring/airbrake_deployment.py validate-modules:doc-missing-type
```
I guess you can also remove this line.

phumpal replied:
Done

phumpal commented:
Here's an updated example using both v2 and v4 deploys apis

```

- hosts: localhost

  tasks:
    - name: notify airbrake about deploy (v2)
      airbrake_deployment:
        repo: https://github.com/phumpal/ansible
        token: '{{ lookup("env", "AIRBRAKE_PROJECT_KEY") }}'
        environment: 'production'
        user: '{{ lookup("env", "DEPLOY_USER") or lookup("env", "USER") }}'
        url: 'https://api.airbrake.io/deploys.txt'
        revision: '012345'

    - name: notify airbrake about deploy (v4)
      airbrake_deployment:
        repo: https://github.com/phumpal/ansible
        project_id: '{{ lookup("env", "AIRBRAKE_PROJECT_ID") }}'
        project_key: '{{ lookup("env", "AIRBRAKE_PROJECT_KEY") }}'
        environment: 'production'
        url: 'https://api.airbrake.io/api/v4/projects/'
        user: '{{ lookup("env", "DEPLOY_USER") or lookup("env", "USER") }}'
        revision: 'f06196ff4a'

```

felixfontein commented on changelogs/fragments/124-airbrake_deployments-api_v4_for_deploy_notices.yml:
```diff
@@ -0,0 +1,6 @@
+- "airbrake_deployment - Allow deploy notifications for Airbrake compatible v2 api (e.g. Errbit)"
+
+minor_changes:
+- "airbrake_deployment - Add deprecation notice for ``token`` parameter and v2 api deploys"
```
This should be in a `deprecated_features:` section. Please also add that this feature will be removed in `Ansible 2.14`. (We'll eventually adjust that string, but right now it's good enough for search+replace :) )

phumpal replied:
done

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -78,44 +98,80 @@ def main():
     module = AnsibleModule(
         argument_spec=dict(
-            token=dict(required=True, no_log=True),
+            token=dict(required=False),
```
```suggestion
            token=dict(required=False, no_log=True, removed_in_version='2.14'),
```
I think the `no_log=True` needs to stay there.

Also, I've just remembered that we had an easier way to deprecate parameters than having to manually call `module.deprecate()`: just add `removed_in_version='2.14'`. I would still leave the manual `module.deprecate()` in the code though, since it explicitly mentions alternatives.

phumpal commented:
>  I assume you tested it both with the v2 and v4

Yep, if you can export env vars for your project you could use https://github.com/ansible-collections/community.general/pull/124#issuecomment-618035027

```
 __________________
< PLAY [localhost] >
 ------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 ________________________
< TASK [Gathering Facts] >
 ------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

Thursday 23 April 2020  01:35:38 -0500 (0:00:00.086)       0:00:00.086 ********
ok: [localhost]
 __________________________________________
< TASK [notify airbrake about deploy (v2)] >
 ------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

Thursday 23 April 2020  01:35:38 -0500 (0:00:00.671)       0:00:00.758 ********
[DEPRECATION WARNING]: Parameter 'token' is deprecated in 2.10. Please remove it and use 'project_id' and
'project_key' instead. This feature will be removed in version 2.14. Deprecation warnings can be disabled by setting
deprecation_warnings=False in ansible.cfg.
changed: [localhost]
 __________________________________________
< TASK [notify airbrake about deploy (v4)] >
 ------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

Thursday 23 April 2020  01:35:39 -0500 (0:00:01.220)       0:00:01.979 ********
changed: [localhost]
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

localhost                  : ok=3    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

Thursday 23 April 2020  01:35:40 -0500 (0:00:00.707)       0:00:02.686 ********
===============================================================================
notify airbrake about deploy (v2) ----------------------------------------------------------------------------- 1.22s
notify airbrake about deploy (v4) ----------------------------------------------------------------------------- 0.71s
Gathering Facts ----------------------------------------------------------------------------------------------- 0.67s
```

> I can't test it

If you open an account support can probably help you with that.

phumpal commented:
Thanks for your help @felixfontein!

felixfontein commented:
> > I can't test it
> 
> If you open an account support can probably help you with that.

Thanks, but I'm not that interested in this module (and there are a million other things I'm busy with ;) ).

phumpal commented:
Ping @Akasurde PTAL again.

felixfontein commented:
@Akasurde mind if I merge this?

felixfontein commented:
@phumpal can you rebase? If tests pass, I'll merge then.

phumpal commented:
@felixfontein I can certainly try. Thanks for the ping.

felixfontein commented:
@phumpal it looks like you didn't rebase against the newest version of `master`, as there's still a conflict in `tests/sanity/ignore-2.10.txt`.

felixfontein commented on plugins/modules/monitoring/airbrake_deployment.py:
```diff
@@ -74,44 +94,80 @@ def main():
+            params["deploy[scm_revision]"] = module.params["revision"]
 
-    url = module.params.get('url')
+        module.deprecate("Parameter 'token' is deprecated in 2.10. Please remove it and use 'project_id' and 'project_key' instead", version='2.12')
```
Please keep 2.14 instead of 2.12. If the sanity tests complain, add a ignore.txt entry for it.

felixfontein commented on tests/sanity/ignore-2.9.txt:
```diff
@@ -699,6 +699,7 @@ plugins/modules/identity/keycloak/keycloak_clienttemplate.py validate-modules:do
 plugins/modules/identity/keycloak/keycloak_clienttemplate.py validate-modules:parameter-type-not-in-doc
 plugins/modules/identity/opendj/opendj_backendprop.py validate-modules:doc-missing-type
 plugins/modules/identity/opendj/opendj_backendprop.py validate-modules:parameter-type-not-in-doc
+plugins/modules/monitoring/airbrake_deployment.py pylint:collection-invalid-deprecated-version
```
No need to add an entry to `ignore-2.9.txt`, since this is a 2.10-only test that fails.

felixfontein commented:
@phumpal thanks for improving the module!

phumpal commented:
Thanks again for your help @felixfontein !
